src/list.hpp

```diff
@@ -204,16 +207,57 @@ void List::set(Context& ctx, size_t row_ndx, T&& value, bool update)
     dispatch([&](auto t) { this->set(row_ndx, ctx.template unbox<std::decay_t<decltype(*t)>>(value, true, update)); });
 }
 
+template <class T>
+inline bool help_compare_values(const T& v1, const T& v2)
+{
+    return v1 != v2;
+}
+
+template <>
+inline bool help_compare_values(const RowExpr& v1, const RowExpr& v2)
+{
+    return v1.get_table() != v2.get_table() || v1.get_index() != v2.get_index();
+}
+
+
+template<typename T, typename Context>
+void List::set_if_different(Context& ctx, size_t row_ndx, T&& value, bool update)
+{
+    dispatch([&](auto t) {
+        using U = std::decay_t<decltype(*t)>;
+        auto old_value =  this->get<U>(row_ndx);
+        auto new_value = ctx.template unbox<U>(value, true, update, true);
+        if (help_compare_values(old_value, new_value))
+            this->set(row_ndx, new_value);
+    });
+}
+
+
 template<typename T, typename Context>
-void List::assign(Context& ctx, T&& values, bool update)
+void List::assign(Context& ctx, T&& values, bool update, bool update_only_diff)
 {
     if (ctx.is_same_list(*this, values))
         return;
 
-    remove_all();
-    if (ctx.is_null(values))
+    if (ctx.is_null(values)) {
+        remove_all();
         return;
+    }
+
+    if (update_only_diff) {
+        size_t sz = util::any_cast<std::vector<util::Any>&>(values).size();
```

Comment:
The concrete type of `values` is not `std::vector<util::Any>` in non-test code. The values passed around have to always be treated as opaque types which are only ever touched by passing them to context functions. The size may also not be known in advance, as we support assigning a non-container enumerator to a list. Fortunately we don't actually need the size in advance, and can do something like the following:

```
size_t sz = size();
size_t index = 0;
ctx.enumerate_list(values, [&](auto&& element) {
    if (index < sz)
        this->set_if_different(ctx, index, element, update);
    else
        this->add(ctx, element, update);
    index++;
});
if (index < sz)
    remove(index, sz);
```

Reply:
Good idea. Thanks.